Introduce the Flower Enhancement Docs process (#1122)
adap/flower: proposals/0001-20220311-flower-enhancement-proposals.md

---
ep-number: 0001

> title: Flower Enhancement Proposals

Reviewer: Flower Enhancements Design?
Flower Enhancements Doc?

Reviewer 2: I like "Flower Enhancement Doc"

- a design document

into one file, which is created incrementally in collaboration with the community.

## Motivation

For far-fetching changes or features proposed to Flower, an abstraction beyond a single Github issues or pull request is required in order to understand and communicate upcoming changes to the project.

The purpose of this process is to reduce the amount of "tribal knowledge" in our community. By moving decisions from Slack threads, video calls and hallway conversations into a well tracked artifact, this process aims to enhance communication and discoverability.

### Goals

Roughly any larger, user facing enhancement should follow the Enhancement Proposal process. If an enhancement would be described in either written or verbal communication to anyone besides the author or developer, then consider creating an Enhancement Proposal.

Similarly, any technical effort (refactoring, major architectural change) that will impact a large section of the development community should also be communicated widely. The Enhancement Proposal process is suited for this even if it will have zero impact on the typical user or operator.

### Non-Goals

> For small changes and additions, going through the Enhancement Proposal process would be time-consuming and unnecessary. This, e.g., includes adding new Federated Learning algorithms, as these only add features without changing how flower works or is used.

Reviewer: ```suggestion
For small changes and additions, going through the Enhancement Proposal process would be time-consuming and unnecessary. This includes, for example, adding new Federated Learning algorithms, as these only add features without changing how Flower works or is used.
```

### Enhancement Proposal Template

Each proposal is provided as a Markdown file having the following structure

- Metadata  
  Metadata as [described below](#metadata) in form of a YAML preamble
- Table of Contents (if needed)
- Summary  
- Motivation
  - Goals
  - Non-Goals
- Proposal
  - Notes/Constraints/Caveats (optional)
- Design Details (optional)
  - Graduation Criteria
  - Upgrade/Downgrade Strategy (if applicable)
- Drawbacks
- Alternatives Considered

As reference, this document follows the above structure.

### Metadata

- __ep-number__ (Required)
  The `ep-number` of the last proposal + 1. With this number it becomes easy to reference other proposals.
- __title__ (Required)  
  The title of the proposal in plain language.
- __status__ (Required)  
  The current status of the proposal. See [workflow](#workflow) for the possible states.
- __authors__ (Required)  
  A list of authors of the proposal. This is simply the Github ID.
- __creation-date__ (Required)  
  The date that the proposal was first submitted in a PR.
- __last-updated__ (Optional)
  The date that the proposal was last changed significantly.  
- __see-also__ (Optional)  

>   A list of other proposals that are releavent to this one.

Reviewer: ```suggestion
  A list of other proposals that are relevant to this one.
```

- __replaces__ (Optional)  
  A list of proposals that this one replaces.
- __superseded-by__ (Optional)  
  A list of proposals that this one supersedes.

### Workflow

> The idea forming the enhancement should already have been discussed or pitched in the comminuty. As such, it needs a champion, usually the author, who shepherds the proposal. This person also has to find committers to flower willing to review the proposal.

Reviewer: ```suggestion
The idea forming the enhancement should already have been discussed or pitched in the comminuty. As such, it needs a champion, usually the author, who shepherds the proposal. This person also has to find committers to Flower willing to review the proposal.
```
